front50-s3/src/main/java/com/netflix/spinnaker/front50/model/S3NotificationDAO.java

```diff
@@ -0,0 +1,132 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.front50.model;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.AmazonS3Exception;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spinnaker.front50.exception.NotFoundException;
+import com.netflix.spinnaker.front50.model.notification.HierarchicalLevel;
+import com.netflix.spinnaker.front50.model.notification.Notification;
+import com.netflix.spinnaker.front50.model.notification.NotificationDAO;
+import org.springframework.beans.factory.annotation.Autowired;
+import rx.Scheduler;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class S3NotificationDAO extends S3Support<Notification> implements NotificationDAO {
+  private final AmazonS3 amazonS3;
+  private final String bucket;
+
+  @Autowired
+  private ObjectMapper objectMapper;
+
+  public S3NotificationDAO(AmazonS3 amazonS3,
+                           Scheduler scheduler,
+                           int refreshIntervalMs,
+                           String bucket,
+                           String rootFolder) {
+    super(amazonS3, scheduler, refreshIntervalMs, bucket, (rootFolder + "/application/").replaceAll("//", "/"));
+
+    this.amazonS3 = amazonS3;
+    this.bucket = bucket;
+  }
+
+  @Override
+  public List<Notification> list() {
+    return allItemsCache.get().stream().collect(Collectors.toList());
+  }
+
+  @Override
+  public Notification getGlobal() {
+    return get(HierarchicalLevel.GLOBAL, Notification.GLOBAL_ID);
+  }
+
+  @Override
+  public Notification get(HierarchicalLevel level, String name) {
+    try {
+      S3Object s3Object = amazonS3.getObject(bucket, buildS3Key(name));
+      Notification notification = objectMapper.readValue(s3Object.getObjectContent(), Notification.class);
+      notification.setLastModified(s3Object.getObjectMetadata().getLastModified().getTime());
+      return notification;
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
+    } catch (AmazonS3Exception e) {
+      if (e.getStatusCode() == 404) {
+        return new Notification();
```

Comment:
Comments like these (any many other tidbits throughout this PR) should go in the code. If I were to change this code in the future, I wouldn't know to look for helpful information in Github.